internal/speakerlist/speakerlist.go

```diff
@@ -80,36 +83,107 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
 }
 
 // Start starts the needed goroutines
-func (sl *SpeakerList) Start(client *k8s.Client) error {
+func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.client = client
 
 	if sl.mList == nil {
-		return nil
+		return
 	}
 
 	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
 
-	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
-	if err != nil {
-		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
-		return err
+func (sl *SpeakerList) memberlistJoin() {
+	if sl.mList == nil {
+		return
+	}
+	joinList := sl.joinList()
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) joinList() []string {
+	joinList := []string{}
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return joinList
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	for name, ip := range sl.k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	return joinList
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.memberlistJoin()
+		}
 	}
-	n, err := sl.mList.Join(iplist)
-	sl.l.Log("op", "startup", "msg", "Memberlist join", "nb joigned", n, "error ?", err)
+}
 
-	return nil
+// SetSpeakersK8S update the list of speakers from the K8S API point of view
+func (sl *SpeakerList) SetSpeakersK8S(eps *v1.Endpoints) {
+	newmap := map[string]string{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			newmap[*ep.NodeName] = ep.IP
+		}
+	}
+	sl.Lock()
+	oldmap := sl.k8sSpeakers
+	sl.Unlock()
+	if !reflect.DeepEqual(oldmap, newmap) {
+		sl.l.Log("op", "setSpeakersK8S", "msg", "New K8S Speakers list set", "old-list", oldmap, "new-list", newmap)
+		sl.Lock()
+		sl.k8sSpeakers = newmap
+		sl.Unlock()
+		sl.memberlistJoin()
+		// needed for split brain and disabled MemberList
+		sl.client.ForceSync()
+	}
 }
 
-// UsableSpeakers return a map of usable nodes
-func (sl *SpeakerList) UsableSpeakers() map[string]bool {
-	if sl.mList != nil {
-		activeNodes := map[string]bool{}
-		for _, n := range sl.mList.Members() {
-			activeNodes[n.Name] = true
+// UsableSpeakers return a map of usable nodes or an error
+func (sl *SpeakerList) UsableSpeakers() (map[string]bool, error) {
```

Comment:
Relaying on memberlist to detect node failures and on the k8s apiserver (that is slow) to add them, seems tricky. We have two source of truths now (instead of just one), so we need to combine them with care.

The only reason I think we are doing this is because memberlist doesn't support to add the nodes again, right? Checking bugs upstream, I see: https://github.com/hashicorp/memberlist/issues/140 and https://github.com/hashicorp/memberlist/issues/80#issuecomment-223316449. They mention, basically, that it might make sense to do it at a higher layer (like serf). I'm not familiar with them to know if it makes sense to use serf or just memberlist, do you? To me, these hacks seem like an indicator that maybe we need to reconsider (maybe after reconsideration, we understand the trade-offs and decide to go with memberlist, though).

What do you think?

Reply:
I'm not sure Serf provide the magic we need https://github.com/hashicorp/serf/issues/333 https://github.com/hashicorp/serf/issues/61 https://github.com/hashicorp/serf/search?q=rejoin&type=Issues

Reply:
I see serf has mdns (which we can't use) to reconnect
https://github.com/hashicorp/serf/blob/master/cmd/serf/command/agent/mdns.go

Reply:
The more I look at Serf the less I think Serf is for us if we are just doing liveness. It adds some nice things that we don't need like Lamport clock, but MemberList and Serf are just not built with the idea of having another source of truth as Consul, so we need to implement the missing features ourselves.